### PR TITLE
feat: ike install-operator command for deploying the operator

### LINKED ISSUE #126: Installer - provide a command to deploy operator
##### What can be improved

By providing `ike install` command we can simplify the installation process for the end-user which will hide all the complexity.

We could bundle processed templates during the build of the ike binary for that purpose. The only thing here to think through would be our local development flow where we change the location of the docker images to internal cluster registry, but I guess we could simply rely on env variables and read them while running `ike install`.

##### Why

The operator installation process involves `Makefile` and several steps one has to perform before using `ike develop`. This is not an appealing option for regular users, especially considering the fact that we aim to ship simple binary. No one will be eager to clone the repo to run `Makefile`.

### PR DESCRIPTION
**Pull request by bartoszmajsak**
#### Short description of what this resolves:

Provides an easy way of installing istio-workspace operator into the cluster through the self-contained binary.

#### Changes proposed in this pull request:

- `ike install` which deploys operator to the cluster
- ability to specify namespace by using `--namespace` or `-n` flag
- tests installer logic in e2e test setup
- bundles deployment resources into the binary by using `go-bindata` library

Fixes #126

### CODE REVIEW

**Comment by bartoszmajsak** [comment]
I am checking what's up with the build, as it obviously works on my machine.

**Comment by bartoszmajsak** [comment]
@aslakknutsen I am just wondering if `install` is not misleading as we have two sides to cover - cluster (which this command provides) and client which requires TP to be present. 

Should we rename to `install-cluster` or provide a flag `--cluster` for later automation of the client-side prerequisites? WDYT?

**Comment by aslakknutsen** [comment]
@bartoszmajsak install-operator? Did you sleep on the "install" vs "dump"?

**Comment by bartoszmajsak** [comment]
I think `install-operator` is fine for the time being.

File: pkg/assets/isto-workspace-deploy.go [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+	names := make([]string, 0, len(_bindata))
+	for name := range _bindata {
+		names = append(names, name)
+	}
+	return names
+}
+
+// _bindata is a table, holding each asset generator, mapped to its name.
+var _bindata = map[string]func() (*asset, error){
+	"deploy/istio-workspace/crds/istio_v1alpha1_session_cr.yaml":  deployIstioWorkspaceCrdsIstio_v1alpha1_session_crYaml,
```
Comment (aslakknutsen) [inline pkg/assets/isto-workspace-deploy.go]: This one is not needed and technically only for testing/example

**Comment by aslakknutsen** [comment]
@bartoszmajsak Some form of feedback on what is going on when you execute the command would be helpful. Print each resource being created or similar. Just nothing makes you wonder if it failed or not.

File: pkg/assets/isto-workspace-deploy.go [hunk 2]
```diff
@@ -0,0 +1,365 @@
+// Package assets Code generated by go-bindata. (@generated) DO NOT EDIT.
```
Comment (aslakknutsen) [inline pkg/assets/isto-workspace-deploy.go]: What's the difference between this and "pkg/assets/generated/isto-workspace-deploy.go"? They seem pretty dup

File: pkg/openshift/client.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
+package openshift
```
Comment (aslakknutsen) [inline pkg/openshift/client.go]: Any reason for keeping this separate from cmd/ike/session/session_client?

File: cmd/ike/cmd/install.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+func installOperator(cmd *cobra.Command, args []string) error { //nolint[:unparam]
+	namespace, err := cmd.Flags().GetString("namespace")
+	if err != nil {
+		return err
+	}
+	app, err := newApplier(namespace)
+	if err != nil {
+		return err
+	}
+	if err := app.applyResource("crds/istio_v1alpha1_session_crd.yaml"); err != nil {
```
Comment (aslakknutsen) [inline cmd/ike/cmd/install.go]: Technically all these could have been replaced with something like:

```go
	resources := []string{"crds/istio_v1alpha1_session_crd.yaml", "service_account.yaml"}
	for _, resource := range resources {
		if r, err := app.applyResource(resource); err != nil { // r = TypeMeta?
			if !errors.IsAlreadyExists(err) {
				log.Error(err, "failed creating " + r)
				return err
			}
			log.Info(r + " already exists")
		}
	}
```

File: pkg/openshift/client.go (hunk 3 above)
Comment (bartoszmajsak) [reply pkg/openshift/client.go]: We can put this together. My thinking was that this one is a generic-purpose client which has no typed interfaces so `openshift` seemed to be a more natural place.

File: pkg/openshift/template.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
+package openshift
```
Comment (aslakknutsen) [inline pkg/openshift/template.go]: Should we separate between "pkg/openshift" the Operator stuff and the "pkg/openshift" 'Other stuff'?

Comment (aslakknutsen) [reply pkg/openshift/template.go]: maybe pkg/openshift/client ?

Comment (aslakknutsen) [reply pkg/openshift/template.go]: or move istio|k8s|openshift "operator stuff" into a common parent package..

File: pkg/openshift/template.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+)
+
+var re = regexp.MustCompile(`(?m)\${(.*?)}`)
+
+const substitution = "{{ .$1 }}"
+
+// ProcessTemplateUsingEnvVars takes template path, loads its content and
+// substitutes variables by using those which are defined as environment variables.
+// Returns processed template as byte array
+func ProcessTemplateUsingEnvVars(templatePath string) ([]byte, error) {
```
Comment (aslakknutsen) [inline pkg/openshift/template.go]: Would have been nice if maybe as part of 'ike install-operator --help' it displayed which ENV variables were available to manipulate. A little dump of the Template.parameters. (note: there is a "description" field on the parameters as well we could use=

Comment (bartoszmajsak) [reply pkg/openshift/template.go]: Yeah, I'm working on it as part of #142

File: cmd/ike/cmd/install.go (hunk 4 above)
Comment (aslakknutsen) [reply cmd/ike/cmd/install.go]: as I wrote that I do now see there are applyResource and applyTemplate but.. :)

File: pkg/assets/isto-workspace-deploy.go (hunk 1 above)
Comment (bartoszmajsak) [reply pkg/assets/isto-workspace-deploy.go]: Renamed file to `istio_v1alpha1_session_cr.example.yaml` and made `go-bindata` ignoring all example files. See 1616034

File: pkg/assets/isto-workspace-deploy.go (hunk 2 above)
Comment (bartoszmajsak) [reply pkg/assets/isto-workspace-deploy.go]: Seems to me like leftover before tweaking go-bindata Make target. Fixed 72ac2c3

File: cmd/ike/cmd/install.go (hunk 4 above)
Comment (bartoszmajsak) [reply cmd/ike/cmd/install.go]: True, I felt this code is crap :) Fixed in 99a78b5

File: pkg/openshift/client.go (hunk 3 above)
Comment (aslakknutsen) [reply pkg/openshift/client.go]: Was not thinking about the location, but rather function. Minimally we could use the 'non typed' 'client' internally within the 'SessionClient' interface. It's our Resource so we shouldn't really be surprised by changes to it. hehe

**Comment by bartoszmajsak** [comment]
> @bartoszmajsak Some form of feedback on what is going on when you execute the command would be helpful. Print each resource being created or similar. Just nothing makes you wonder if it failed or not.

Done in db2da47

File: pkg/openshift/client.go (hunk 3 above)
Comment (bartoszmajsak) [reply pkg/openshift/client.go]: Ah ok, got it. Let me go over the code and see - there's definitely some dupl in a way the client is set up for example. We can have both typed and generic in one place, why not.

File: pkg/openshift/template.go (hunk 5 above)
Comment (bartoszmajsak) [reply pkg/openshift/template.go]: I did a small shuffling around. I opened #150 issue to clean up the client stuff.

File: pkg/openshift/client.go (hunk 3 above)
Comment (bartoszmajsak) [reply pkg/openshift/client.go]: I opened #150 issue to clean up the client stuff.

**Pull request merged**